crates/cheatcodes/src/toml.rs

```diff
@@ -0,0 +1,214 @@
+    }
+}
+
+impl Cheatcode for parseTomlStringArrayCall {
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        let Self { toml, key } = self;
+        parse_json_coerce(&convert(toml)?, key, &DynSolType::String)
+    }
+}
+
+impl Cheatcode for parseTomlBytesCall {
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        let Self { toml, key } = self;
+        parse_json_coerce(&convert(toml)?, key, &DynSolType::Bytes)
+    }
+}
+
+impl Cheatcode for parseTomlBytesArrayCall {
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        let Self { toml, key } = self;
+        parse_json_coerce(&convert(toml)?, key, &DynSolType::Bytes)
+    }
+}
+
+impl Cheatcode for parseTomlBytes32Call {
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        let Self { toml, key } = self;
+        parse_json_coerce(&convert(toml)?, key, &DynSolType::FixedBytes(32))
+    }
+}
+
+impl Cheatcode for parseTomlBytes32ArrayCall {
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        let Self { toml, key } = self;
+        parse_json_coerce(&convert(toml)?, key, &DynSolType::FixedBytes(32))
+    }
+}
+
+impl Cheatcode for parseTomlKeysCall {
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        let Self { toml, key } = self;
+        parse_json_keys(&convert(toml)?, key)
+    }
+}
+
+impl Cheatcode for writeToml_0Call {
+    fn apply(&self, state: &mut Cheatcodes) -> Result {
+        let Self { json, path } = self;
+        let value =
+            serde_json::from_str(json).unwrap_or_else(|_| JsonValue::String(json.to_owned()));
+        let toml = json_to_toml(value);
+        let toml_string = toml::to_string_pretty(&toml).expect("failed to serialize TOML");
+        super::fs::write_file(state, path.as_ref(), toml_string.as_bytes())
+    }
+}
+
+impl Cheatcode for writeToml_1Call {
+    fn apply(&self, state: &mut Cheatcodes) -> Result {
+        let Self { json, path, valueKey } = self;
+        let json =
+            serde_json::from_str(json).unwrap_or_else(|_| JsonValue::String(json.to_owned()));
+
+        let data_path = state.config.ensure_path_allowed(path, FsAccessKind::Read)?;
+        let toml_data = fs::read_to_string(data_path)?;
+        let json_data = toml_to_json(parse_toml_str(&toml_data)?);
+        let value =
+            jsonpath_lib::replace_with(json_data, &canonicalize_json_path(valueKey), &mut |_| {
+                Some(json.clone())
+            })?;
+
+        let toml = json_to_toml(value);
+        let toml_string = toml::to_string_pretty(&toml).expect("failed to serialize TOML");
+        super::fs::write_file(state, path.as_ref(), toml_string.as_bytes())
+    }
+}
+
+fn parse_toml_str(toml: &str) -> Result<TomlValue> {
+    toml::from_str(toml).map_err(|e| fmt_err!("failed parsing TOML: {e}"))
+}
+
+/// Convert a TOML string to a JSON string.
+fn convert(toml: &str) -> Result<String> {
+    let toml = parse_toml_str(toml)?;
+    let json = toml_to_json(toml);
```

Comment:
Would it be possible to avoid this conversion by just using the `Serialize` implementation on `TomlValue`? https://docs.rs/toml/latest/src/toml/value.rs.html#398-443

Reply:
Good point! I initially ran into issues with the direct conversion, specifically around arrays and objects, but those turned out to be unrelated. I've updated the implementation to use direct conversions, much nicer.